File: fragment/nodes/drawable.py
from .node2d import Node2D
from .canvas import Canvas
from pygame_render import Texture


class Drawable(Node2D):
    """The base node class for drawing things onto the screen.

    Drawable locates nearest Canvas ancestor (if none found, it uses the global canvas)
    and renders itself there.
    """
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.renderer = self.game_manager.window_manager.renderer
        self.target_canvas = self.find_ancestor_of_type(Canvas) or self.renderer.global_canvas
        self.target_canvas.register_drawable(self)

    def render(self) -> Texture:
        """Renders the drawable onto render layer."""
        texture = self.renderer.placeholder_layer.texture
        return texture

    def destroy_self(self) -> None:
        self.target_canvas.unregister_drawable(self)
        super().destroy_self()

    @property
    def draw_priority(self) -> int:
        return self.properties['draw_priority']

    @draw_priority.setter
    def draw_priority(self, priority: int) -> None:
        self.properties['draw_priority'] = int(priority)

File: fragment/nodes/drawable.js
import { Node2D } from '/fragment/nodes/node2d.js';
import { Canvas } from '/fragment/nodes/canvas.js';

export class Drawable extends Node2D {
    /**
     * Drawable locates nearest Canvas ancestor (if none found, it uses the global canvas)
     * and renders itself there.
     */
    constructor(...args) {
        super(...args);
        this.pixiSprite = new PIXI.Sprite();
        this.pixiSprite.anchor.set(0.5, 0.5); // Center anchor
    }

    fullInit() {
        super.fullInit();
        this.targetCanvas = null;
        this.updateTargetCanvas();
    }

    update(dt) {
        super.update(dt);
        this.updatePixiTransform();
    }

    updatePixiTransform() {
        if (this.destroyed) return;
        this.pixiSprite.position.set(this.worldPosition.x, this.worldPosition.y);
        this.pixiSprite.rotation = this.worldRotation * Math.PI / 180;
        this.pixiSprite.scale.set(this.worldScale.x, this.worldScale.y);
    }

    updateTargetCanvas() {
        let target = this.findAncestorOfType(Canvas)
        if (target === null || target.destroyed) {
            target = this.gameManager.windowManager.globalCanvas;
        }
        if (this.targetCanvas === target) {
            return;
        }
        if (this.targetCanvas !== null && !this.targetCanvas.destroyed) {
            this.targetCanvas.unregisterDrawable(this);
        }
        this.targetCanvas = target;
        this.targetCanvas.registerDrawable(this);
    }

    destroySelf() {
        if (!this.targetCanvas.destroyed) {
            this.targetCanvas.unregisterDrawable(this);
        }
        this.pixiSprite.destroy({ texture: true });
        super.destroySelf();
    }

    get drawPriority() {
        return this._properties['drawPriority'];
    }

    set drawPriority(priority) {
        this._properties['drawPriority'] = parseInt(priority);
        this.pixiSprite.zIndex = this._properties['drawPriority'];
    }
}
